pkg/secretmanager/provide.go

```diff
@@ -0,0 +1,81 @@
+package secretmanager
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	agentinfo "github.com/fluxninja/aperture/v2/pkg/agent-info"
+	"github.com/fluxninja/aperture/v2/pkg/utils"
+	"go.uber.org/fx"
+	"golang.org/x/oauth2/google"
+)
+
+// Module is a fx module that provides secret manager instance.
+func Module() fx.Option {
+	return fx.Options(
+		fx.Provide(
+			provideSecretManagerClient,
+		),
+	)
+}
+
+// SecretManagerClientIn holds parameters for provideSecretManagerClient.
+type SecretManagerClientIn struct {
+	fx.In
+	Shutdowner             fx.Shutdowner
+	Lifecycle              fx.Lifecycle
+	InstallationModeConfig *agentinfo.InstallationModeConfig `optional:"true"`
+}
+
+// SecretManagerClient is a wrapper around secretmanager.Client.
+type SecretManagerClient struct {
+	Client       *secretmanager.Client
+	GCPProjectID string
+	ProjectID    string
+}
+
+func provideSecretManagerClient(in SecretManagerClientIn) (*SecretManagerClient, error) {
+	if in.InstallationModeConfig.InstallationMode != utils.InstallationModeCloudAgent {
+		return nil, nil
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	client, err := secretmanager.NewClient(ctx)
+	if err != nil {
+		defer cancel()
```

Comment:
If we defer cancel (see above whether we should) we probably want to do this once anyway?

Reply:
Here, the cancel has to be called either in OnStop when the agent is stopping or when there is some errors. We can't defer it once as it would cancel the context even before the onStart is executed and we would be using cancelled context after OnStart